cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll

```diff
@@ -175,6 +176,23 @@ private predicate nodeIsBarrier(DataFlow::Node node) {
     readsVariable(node.asInstruction(), checkedVar) and
     hasUpperBoundsCheck(checkedVar)
   )
+  or
+  exists(Variable checkedVar, IRGuardCondition guard, Operand access, Operand other |
+    /*
+     * This node is guarded by a condition that forces the accessed variable
+     * to equal something else.  For example:
+     * ```
+     * x = taintsource()
+     * if (x == 10) {
+     *   taintsink(x); // not considered tainted
+     * }
+     * ```
+     */
+
+    readsVariable(node.asInstruction(), checkedVar) and
+    readsVariable(access.getDef(), checkedVar) and
```

Comment:
How is performance? Are the tuple counts free of bulges?

Reply:
Your intuition was correct - it joined the two `readsVariable`s first leading to a bit of a tuple explosion (though performance was still OK on the two snapshots I tried).  I've fixed this by pulling some of the logic out into a separate predicate.